### ISSUE TITLE
[lua] Provide information about number of instructions left

### ISSUE THREAD

**Issue by azaz44** [FrSkyRC/ETHOS-Feedback-Community#4103]
This limit was introduced in Ethos 1.5.3:

> [lua] Maximum instructions limit added (to avoid infinite loops)

which cause a lot of scripts to fail with this or similar error:
```
007973 wakeup() error: /scripts/Flog/main.lua:1830: Max instructions count reached
```

This is a major problem for all long-time processing operations which need to span across multiple `wakeup` cycles. For example reading / parsing a log file, reading big configuration files etc.

I think, if we have a limited number of instructions, we also need some way to query how much is still left. Something like:
```lua
system.getAvailableInstructionCount()
```

This would allow to check what is still available and decide, if we can do more processing in this `wakeup` cycle, or not.

We had a similar limit in OpenTX, which was based on "CPU Usage" and was killing a script if it made too much work in one cycle. But the API was providing `getUsage()` function which allowed the script to query current value and make decisions. It wasn't perfect but it worked.

I think "instruction limit" is better (than CPU usage) because it's more predictable, but we need some API help.

I would actually love if this limit was temporarily removed, until we get some solution. A lot of scripts will need to be modified after introduction of this limit, but at the moment there's not even any way do do this.

---
Besides that, I think it would be great, if max instruction limit was bigger for initialization functions (`init`, `create`) than it is for `wakeup`, `paint`, `event`. This would allow scripts to initialize (which sometimes involves reading files) in a much better way. Developing anything to do work over multiple cycles is just so more complex.

Again, I believe it was same in OpenTX, where `init` could do way more work than `run`.

---

Or if there's any solution to query sth right not I would love to know. I checked API and also looked for something in LUA, but I can't find anything.

**Comment by robthomson** [comment]
So from my understanding the only time this occurs is if you have something like an infinite while loop.

if you write your code to use wakeup correctly - then you never have this issue.

Not saying the function does not have a purpose.. but I dont think the limit is of any impact if you write the code with it in consideration.

**Comment by azaz44** [comment]
> So from my understanding the only time this occurs is if you have something like an infinite while loop.
> 
> if you write your code to use wakeup correctly - then you never have this issue.
> 
> Not saying the function does not have a purpose.. but I dont think the limit is of any impact if you write the code with it in consideration.

So how would you consider this? Let's assume we want to read the log file which might be few MB big.

**Comment by robthomson** [comment]
I personally would have logged to multiple smaller files - avoiding the issue on the first place.

But you could chunk the read on each loop - starting at a different location in the file per read.... Gradually building up the file in ram.

Not ideal.. But would work.

End of day solution is prob to avoid big files in first place!   (IMHO)

**Comment by azaz44** [comment]
> I personally would have logged to multiple smaller files - avoiding the issue on the first place.

You still need to read them.
And you might be reading files which were written - for example - by Ethos or somebody else. Processing a file was just an example of any long running operation. 

> But you could chunk the read on each loop - starting at a different location in the file per read.... Gradually building up the file in ram.

That's exactly how it is supposed to work. But for this you need sth to tell you, if you can still read one line in this wakeup cycle, or if you need to defer further processing to the next cycle.

You don't want to process to little in one wakeup, for example read just one line, becuse processing will take ages. But at the same time you can't exceed the limit.

**Comment by azaz44** [comment]
**TL;DR version: Can we please have this limit disabled until better solution is found. This will cause a lot of hard-to-find problems with many LUA scripts**

There is one more problem with this limit. It allows the script to continue, which causes a lot of problems.

For example, I have a script which loads a log of recent flights. It can show statistics, and a list of recent flights grouped by day, among others. To give an idea:

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/42830563/a4505641-0f32-4fc6-9409-8dc277d65671)

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/42830563/61955ea4-15ad-4dd0-b9e8-4fbaf6a693d5)

This is a "system tool". The CSV file is updated by another script (registerSource) which detects flight end and adds an entry to the file.

Now the problem is, I'm getting wrong data, missing flights, wrong statistics, all kind of problems. As you see, it reports 10 flights done on 29th, while CSV has 12

```
2024-05-27,19:43:26,2024-05-27,19:50:16,Kraken 700,399,B14+B15,Time Left=9s;mAh Consumed=4280mAh;RSSI Min=42dB;VFR Min=75%
2024-05-27,19:55:31,2024-05-27,20:06:24,Kraken 700,357,B11+B13,Time Left=0s;mAh Consumed=3929mAh;RSSI Min=42dB;VFR Min=80%
2024-05-27,20:19:06,2024-05-27,20:25:08,Kraken 700,363,B07+B08,Time Left=0s;mAh Consumed=3932mAh;RSSI Min=36dB;VFR Min=82%
2024-05-27,20:30:48,2024-05-27,20:36:50,Kraken 700,363,B09+B10,Time Left=3s;mAh Consumed=3861mAh;RSSI Min=43dB;VFR Min=78%
2024-05-29,18:11:02,2024-05-29,18:21:26,Kraken 700,562,B04+B12,Time Left=0s;mAh Consumed=4506mAh;RSSI Min=38dB;VFR Min=81%
2024-05-29,18:26:17,2024-05-29,18:31:42,Kraken 700,326,B01+B02,Time Left=7s;mAh Consumed=3744mAh;RSSI Min=38dB;VFR Min=86%
2024-05-29,18:37:57,2024-05-29,18:45:13,Kraken 700,427,B14+B15,Time Left=0s;mAh Consumed=4438mAh;RSSI Min=43dB;VFR Min=82%
2024-05-29,18:50:05,2024-05-29,18:57:21,Kraken 700,426,B11+B13,Time Left=4s;mAh Consumed=3850mAh;RSSI Min=38dB;VFR Min=86%
2024-05-29,19:04:33,2024-05-29,19:10:57,Kraken 700,369,B07+B08,Time Left=0s;mAh Consumed=3996mAh;RSSI Min=43dB;VFR Min=84%
2024-05-29,19:15:55,2024-05-29,19:22:20,Kraken 700,371,B09+B10,Time Left=0s;mAh Consumed=3893mAh;RSSI Min=43dB;VFR Min=87%
2024-05-29,19:34:44,2024-05-29,19:45:18,Kraken 700,590,B04+B12,Time Left=9s;mAh Consumed=4387mAh;RSSI Min=42dB;VFR Min=85%
2024-05-29,19:51:17,2024-05-29,19:56:55,Kraken 700,339,B01+B02,Time Left=11s;mAh Consumed=3701mAh;RSSI Min=43dB;VFR Min=82%
2024-05-29,20:04:34,2024-05-29,20:10:25,Kraken 700,342,B14+B15,Time Left=2s;mAh Consumed=4368mAh;RSSI Min=40dB;VFR Min=94%
2024-05-29,20:16:04,2024-05-29,20:26:36,Kraken 700,633,B11+B13,Time Left=13s;mAh Consumed=3725mAh;RSSI Min=42dB;VFR Min=85%
2024-05-29,20:32:32,2024-05-29,20:38:34,Kraken 700,364,B07+B08,Time Left=2s;mAh Consumed=3868mAh;RSSI Min=38dB;VFR Min=82%
2024-05-29,20:43:35,2024-05-29,20:52:27,Kraken 700,533,B09+B10,Time Left=4s;mAh Consumed=3850mAh;RSSI Min=38dB;VFR Min=87%
```

And everything else is wrong too, list of flights, statistics, usages of models and batteries etc.

This is because when instruction limit is exceeded, something isn't processed. But Ethos silently ignores the problem and invokes another `wakeup` for this script in next cycle. User isn't aware of any error, script has no chance to react or even know about the issue. The error (instruction limit exceeded) is only visible if you connect a serial debugging to the ratio and debug.

I'll just attempt to workaround the issue, will just read less lines in one wakeup cycle (I think I do 5).

But please imagine, there is no way to fix this problem in any secure way. You will never know when your instruction limit is exceeded, and in such case the Ethos terminates this `wakeup`, but continues invoking the next one. This can cause bad data like in my case, but also all kinds of problems. This is well known in software to cause terrible issues and is avoided like a plague ("Swallowing exceptions" we call it).

I think this is a major problem. I would really love to have this limit disabled for now.

----

Also, to consider how it was done in OpenTX. There, if script exceeded the limit (which it could check with `getUsage`), the complete script would be killed. The `run` function (similar to `wakeup`) would never be called again and in addition it would be visible in the UI, that script has an error. I think this is a much safer approach which I would love to see in Ethos.

Killing a single wakeup cycle with unfinished work and then invoking another one doesn't make any sense and is so dangerous.

**Comment by azaz44** [comment]
One more update.

In OpenTX, any script error would cause its termination, and subsequent functions (`run`, `paint`, `event`) would not be called anymore. Not only exceeding script processing time, but any other error too. I would love to see Ethos do same thing, because it is much much safer.

**Comment by azaz44** [comment]
One more update: this thing has been greatly reworked in EdgeTx comparing to OpenTx, just in case it is interesting to anybody.

**Comment by azaz44** [comment]
One more thing:

the limit in Ethos is very restrictive. I can do way less in one script comparing to what I could in OpenTX. I think it is way too low.

And also, in EdgeTX they improved it greatly comparing to OpenTX. Instead of having a limit, they preempt the script and continue later. I believe the PR which introduced it is this one: https://github.com/EdgeTX/edgetx/pull/285 . I would love to see sth similar in Ethos as it is a much better solution.

But we need sth quick for now, just disabling this limit or making it bigger by some big number. I just can't imagine how many scripts which do more then just display some simple widget have issues now.

**Comment by bsongis-frsky** [comment]
I have to decide which function will get preemption enabled. Perhaps only `wakeup()`?
Also I think EdgeTX doesn't allow at all preemption in widgets.

**Comment by azaz44** [comment]
> I have to decide which function will get preemption enabled. Perhaps only `wakeup()`? Also I think EdgeTX doesn't allow at all preemption in widgets.

I would think `wakeup` and initialization at minimum. But in general as much as possible, just in case. So for:
- `registerSystemTool`:
  - `create`, `wakeup`
  - nice to have: `read` and `write` (if ever supported)
  - nice to have: `close` (gives possibility to write some data to a file, save the state, without worrying about exceeding the limit)
  - nice to have: `event` (if possible)
- `registerSource`:
  - `init` (the one specified in the table, returned by init), `wakeup`
![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/42830563/98b72162-bc07-4f55-9db9-85d6a7938d83)
  - nice to have: `read`, `write`, `configure`
- `registerTask` - I didn't test this one yet, but probably
  - `init`, `wakeup`
  - nice to have: `read`, `write` (if ever supported), `configure`
  - nice to have: `event`
- `registerWidget`
  - `create`, `wakeup`
  - nice to have: `read`, `write`, `configure`
  - nice to have: `event`

I assume `paint` might be impossible.
And `event` might be problematic too? I think this because Ethos needs a return value from this function to decide what to do with the event. Preempting this would be great as a lot of processing might happen here. But this can be workarounded by delegating work to `wakeup` if preempting `event` isn't an option. Just would be great to have this somewhere in documentation so script writers are aware.

**Comment by bsongis-frsky** [comment]
`create` is problematic, because its result is used in all other methods
I will start with `wakeup`. I have something already working right now.

**Comment by azaz44** [comment]
This sounds great, I didn't expect this being addressed that soon 👍

**Comment by bsongis-frsky** [comment]
It will need to be tested seriously! Only the slow scripts will be preempted. It seems to me that EdgeTX preempts the whole lua system.

**Comment by azaz44** [comment]
> It will need to be tested seriously! Only the slow scripts will be preempted. It seems to me that EdgeTX preempts the whole lua system.

I'm at global 3d until Tuesday but can test after that. My scripts are large so that should be a good test.

In general there's a problem when this change is not in because we either allow scripts to kill the system with infinite loops, or we make many scripts not work by instruction limit. So I think any solution will likely be an improvement. Happy to test.

**Comment by bsongis-frsky** [comment]
It will be nice if I can have a feedback on this, I would like to have the release during the week-end!

**Comment by robthomson** [comment]
New release would be nice.  I have so many bits waiting for 1.5.10

**Comment by azaz44** [comment]
> It will be nice if I can have a feedback on this, I would like to have the release during the week-end!

Let me test.

First impressions - preempting works for my script. I tested doing work for 0.5 sec in each the wakeup cycle (a bit too much), this would immediately fail on last stable versions (instructions limit exceeded), and now it works great. Everything is much faster thanks to this too. I will test more, as this is only one case (wakeup cycles of a system tool script). But it looks very promising 👍

However I see a new problem - rotary wheel doesn't seen to generate any events for LUA. DISP, SYS, MDL - these work. Pressing rotary works. But rotating does not.

I'll try to test more on the weekend and also fly the nightly a bit in simulator.

**Comment by azaz44** [comment]
@bsongis-frsky Is there any way to preview changes introduced in the nightly? I would try to pay attention to what is being changed while testing.

**Comment by azaz44** [comment]
Update: I tested in the SIM too, and it works great. But there is one change - scripts need to be in `C:\Users\xxxxx\.ethosXX` folder instead of simulator folder. But I might have had a bit outdated simulator there. The change is fine if course.

**Comment by bsongis-frsky** [comment]
> @bsongis-frsky Is there any way to preview changes introduced in the nightly? I would try to pay attention to what is being changed while testing.

The release note explains all changes

**Comment by bsongis-frsky** [comment]
> However I see a new problem - rotary wheel doesn't seen to generate any events for LUA. DISP, SYS, MDL - these work. Pressing rotary works. But rotating does not.

In a system tool? Reproductible with my tool form example?

**Comment by azaz44** [comment]
> > However I see a new problem - rotary wheel doesn't seen to generate any events for LUA. DISP, SYS, MDL - these work. Pressing rotary works. But rotating does not.
> 
> In a system tool? Reproductible with my tool form example?

Yes, in a system tool. I will check the example tomorrow or get something minimal to repro.

**Comment by bsongis-frsky** [comment]
It seems to me that the Snake game doesn't receive the rotary encoder events

**Comment by azaz44** [comment]
> It seems to me that the Snake game doesn't receive the rotary encoder events

I just tested with snake and rotary doesn't do anything, should it? (never tried it before).
But then I downgraded to 1.5.9 and it also doesn't work. Could the but be introduced in 1.5.9? I might have used 1.5.8 before testing the nightly.

But with my scripts, rotary works on 1.4.9.

Also one more thing - in 1.4.9 I can control the snake with MDL / SYS / DISP / RTN.
But in the nightly 1.4.10 MDL was exiting the script and going to model settings screens.

**Comment by bsongis-frsky** [comment]
I have written a fix and uploaded new nightlies. Would you do a quick test before I launch the release?

**Comment by azaz44** [comment]
> I have written a fix and uploaded new nightlies. Would you do a quick test before I launch the release?

I gave it a quick test:
- preempting still works great
  - (I wish I could test all other functions besides wakeup and also other script types, but I cannot do this super quick, would need some time. But judging from release notes only wakeup is affected anyway)
- rotary now works fine in my script (in "snake" not, but I'm not sure if it does anything in snake)
- however there's still a problem that in "Snake" MDL quits the script. It seems it makes it so if you start Snake and press MDL right away. However if you press some other key like RTN first, and then press MDL, MDL works fine (goes up).
- in my own script, I was handling RTN and moving back to a previous screen. But in the nightly RTN just exits my script. I would need more time to test this. I don't know if that's some intended change and I need to update my script, or unintentional bug.

**Comment by azaz44** [comment]
I also see some blinking of slider bars when shutting down, which I think I didn't see before. 100% sure I'm not, but I think I see it for the first time with the nightly:
https://youtu.be/6trRWO6d9uQ

**Comment by bsongis** [comment]
If you press MDL when you are on the top of the screen, perhaps you loose and the game exits ;)

**Comment by bsongis-frsky** [comment]
OK. This needs to be checked. I prefer to avoid the regressions. I will release later.

**Comment by Devil-IdT** [comment]
>  I will release later.

Will there also be files for the X20R/RS EU (LBT) in this release?

**Comment by bsongis-frsky** [comment]
On the video, I see you have 4 home screens. If you only keep the one you show on the video, do you reproduce the problem?

**Comment by azaz44** [comment]
> On the video, I see you have 4 home screens. If you only keep the one you show on the video, do you reproduce the problem?

Yes. Cloned the model, removed all screens except of first and problem is still there. But it only happens on this screen so I think it is related to this particular layout.

Actually just made a test. All 4 layouts which show trims and sliders are affected. Independent how many screens there are and on which position this screen is at.

**Comment by bsongis-frsky** [comment]
Would you send me your model file, and the bitmap? I cannot reproduce it ...

**Comment by bsongis-frsky** [comment]
I can reproduce it. I press PWR and in the same time, I move POT3
Not in dark mode!

**Comment by bsongis-frsky** [comment]
Reproductible with 1.5.8
I think it's a bug which is very old!

**Comment by azaz44** [comment]
> Reproductible with 1.5.8 I think it's a bug which is very old!

ok. Unlucky that it happened to me for the first time on 1.5.10, the slider was probably oscillating or sth.

**Comment by bsongis-frsky** [comment]
OK. This problem will be fixed in 1.5.10. It's there since 4 years!

**Comment by azaz44** [comment]
> OK. This problem will be fixed in 1.5.10. It's there since 4 years!

Super weird that nobody ever saw it 🙂
for me, it's probably because in normal usage I'm never at the first screen when powering off. After flight I'm at screen 3. But how it never happend for others, or for me in the model for simulator, I have no idea. But doesn't matter, good to see release being closer 👍

**Comment by fdm225** [comment]
@bsongis-frsky I am fixing a script that generates a QR code from the last logged GPS location and it does a lot of processing when started, enough to trigger the Max instructions count reached message.  I am doing the processing in the wakeup function on version 1.5.15; do I need to do anything to use the preemption that is discussed in this thread or am I misunderstanding how it works?  I assumed that when the instruction count was reached it would preempt the current running script and when wakeup is called again it would pickup from where it left off, is that correct?

```
local function wakeup(widget)
	doHeavyCalc(widget)
```

<img width="1203" alt="image" src="https://github.com/user-attachments/assets/9c0a0e3b-4b31-43b3-aa0a-23552148f6d6">


thanks,
Dave

**Comment by robthomson** [comment]
@fdm225 can you send me details on how the qr code is being made?   I have been looking for something in lua to do this.

**Comment by azaz44** [comment]
> ipt that generates a QR code from the last logged GPS location and it does a lot of processing when started, enough to trigger the Max instructions count reached message. I am doing the processing in the wakeup function on version 1.5.15; do I need to do anything to use the preemption that is discussed in this thread or am I misunderstanding how it works? I assumed that when t

Yes, as long as you do your processing in the `wakeup` you're good. You can do all processing in one go, and it will happen over multiple cycles automatically.

You might still want to split processing to pieces just to have UI updated every some time with some progress, but that's a bonus.

**Comment by fdm225** [comment]
@robthomson this is not my code, just a project I found on [Github](https://github.com/strgaltdel/Ethos-ModelFinder) . Please note that it currently does not work as it does all of the heavy calculations needed for the QR code in the view function causing the "Max instructions exceeded" error.  I am attempting to modify it like what is written in this issue (move it to the wakeup function) but it's currently not working.

@azaz44 I am calling this from the wakeup function and I am getting the Max instructions exceeded error, so either my code is wrong or it's broken in Ethos 1.5.15 on the X20 Pro.  I have attached the modified script to this post.  Do the developers read comments on closed issues?  is there a way to re-open this issue so the developers see this as if this is something that they plan on supporting (and this isn't a code issue on my part) then I assume they would like to know that their fix isn't working.

dave

[Ethos-ModelFinder.zip](https://github.com/user-attachments/files/16932153/Ethos-ModelFinder.zip)

**Comment by azaz44** [comment]
> @azaz44 I am calling this from the wakeup function and I am getting the Max instructions exceeded error, so either my code is wrong or it's broken in Ethos 1.5.15 on the X20 Pro. I have attached the modified script to this post. Do the developers read comments on closed issues? is there a way to re-open this issue so the developers see this as if this is something that they plan on supporting (and this isn't a code issue on my part) then I assume they would like to know that their fix isn't working.
> 
> dave
> 
> [Ethos-ModelFinder.zip](https://github.com/user-attachments/files/16932153/Ethos-ModelFinder.zip)

I remember I tested this and it worked, I used it in my script. But it's not a widget. I looked at you script just to make sure you do this in `wakeup` (as I think only wakeup is preempted) and script looks fine. So I think it's best if you open a new issue, and maybe link this one as related.
